p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,314 @@
+package holepunch
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/holepunch/pb"
+	"github.com/libp2p/go-libp2p/p2p/protocol/identify"
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// TODO Should we have options for these ?
+const (
+	protocol         = "/libp2p/holepunch/1.0.0"
+	maxMsgSize       = 4 * 1024 // 4K
+	holePunchTimeout = 1 * time.Minute
+	dialTimeout      = 10 * time.Second
+	maxRetries       = 5
+	retryWait        = 2 * time.Second
+)
+
+var (
+	log = logging.Logger("p2p-holepunch")
+)
+
+// TODO Find a better name for this protocol.
+// HolePunchService is used to make direct connections with a peer via hole-punching.
+type HolePunchService struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	ids  *identify.IDService
+	host host.Host
+
+	// ensure we shutdown ONLY once
+	closeSync sync.Once
+	refCount  sync.WaitGroup
+}
+
+// NewHolePunchService creates a new service that can be used for hole punching
+func NewHolePunchService(h host.Host, ids *identify.IDService) (*HolePunchService, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	hs := &HolePunchService{ctx: ctx, ctxCancel: cancel, host: h, ids: ids}
+
+	h.SetStreamHandler(protocol, hs.handleNewStream)
+	h.Network().Notify((*netNotifiee)(hs))
+	return hs, nil
+}
+
+// Close closes the Hole Punch Service.
+func (hs *HolePunchService) Close() error {
+	hs.closeSync.Do(func() {
```

Comment:
The cancel func can be called multiple times. I believe that allows you to get rid of this Once here.